common/remote/dafsstream.cpp

```diff
@@ -0,0 +1,1383 @@
+        MemoryBuffer decompressedMetaInfoMb;
+        fastLZDecompressToBuffer(decompressedMetaInfoMb, compressedMetaInfoMb);
+        Owned<IPropertyTree> metaInfoEnvelope = createPTree(decompressedMetaInfoMb);
+        StringBuffer metaInfoSignature;
+        if (metaInfoEnvelope->getProp("signature", metaInfoSignature))
+        {
+#if !defined(_USE_OPENSSL) || !defined(_WIN32)
+            // implies will not be able to use this to communicate with secure dafilesrv's
+#endif
+            MemoryBuffer metaInfoBlob;
+            metaInfoEnvelope->getPropBin("metaInfoBlob", metaInfoBlob);
+
+            metaInfo.setown(createPTree(metaInfoBlob));
+        }
+        else
+            metaInfo.set(metaInfoEnvelope);
+
+        name.set(metaInfo->queryProp("logicalFilename"));
+        if (name.isEmpty())
+            throwDsFsClientExceptionV(DaFsClient_InvalidMetaInfo, "logicalFilename missing");
+
+        IPropertyTree *fileInfo = metaInfo->queryPropTree("FileInfo");
+        if (!fileInfo)
+            throwDsFsClientExceptionV(DaFsClient_InvalidMetaInfo, "FileInfo is missing (logicalFilename=%s", name.get());
+        unsigned metaInfoVersion = metaInfo->getPropInt("version");
+
+        switch (metaInfoVersion)
+        {
+            case 0:
+                // implies unsigned direct request from engines (on unsecure port)
+                // fall through
+            case 1:
+            {
+                // old metaInfo, reconstruct a IFileDescriptor for ease of compatibility with rest of code
+
+                unsigned numParts = fileInfo->getCount("Part");
+
+                // calculate part mask
+                const char *path = fileInfo->queryProp("Part[1]/Copy[1]/@filePath");
+                if (isEmptyString(path))
+                    throwDsFsClientExceptionV(DaFsClient_InvalidMetaInfo, "filePath not found (logicalFilename=%s", name.get());
+                StringBuffer dir, fname, ext;
+                splitFilename(path, &dir, &dir, &fname, &ext);
+                VStringBuffer partMask("%s._$P$_of_%u", fname.str(), numParts);
+
+                // reconstruct group
+                SocketEndpointArray eps;
+                bool replicated = false;
+                Owned<IPropertyTreeIterator> iter = fileInfo->getElements("Part");
+                ForEach(*iter)
+                {
+                    IPropertyTree &part = iter->query();
+                    if (part.hasProp("Copy[2]"))
+                        replicated = true;
+                    const char *host = part.queryProp("Copy[1]/@host");
+                    SocketEndpoint ep(host);
+                    eps.append(ep);
+                }
+                StringBuffer groupText;
+                eps.getText(groupText);
+                Owned<IGroup> group = createIGroup(eps);
+                ClusterPartDiskMapSpec mspec;
+                mspec.defaultCopies = replicated?DFD_DefaultCopies:DFD_NoCopies;
```

Comment:
in the replicated case, shouldn't we get the actual number of copies from the file structure rather than defaulting?

Reply:
yes, but this is legacy (case 1:) where Esp didn't provide enough info in the response.
In the case of 7.2 esp (case 2:), the whole file descriptor is serialized, which contains much the reasl ClusterPartDiskMapSpec + other info.